ports/unix/main.c

```diff
@@ -297,18 +234,28 @@ static int do_repl(void) {
             line = line3;
         }
 
-        int ret = execute_from_lexer(LEX_SRC_STR, line, MP_PARSE_SINGLE_INPUT, true);
+        ret = execute_from_lexer(LEX_SRC_STR, line, MP_PARSE_SINGLE_INPUT, true);
         free(line);
         if (ret & FORCED_EXIT) {
             return ret;
         }
     }
-
     #endif
+    return ret;
 }
 
+
 static int do_file(const char *file) {
-    return execute_from_lexer(LEX_SRC_FILENAME, file, MP_PARSE_FILE_INPUT, false);
+    int ret = pyexec_file(file);
+    // pyexec returns 1 for success, 0 for exception, PYEXEC_FORCED_EXIT for SystemExit
+    // Convert to unix port's expected codes: 0 for success, 1 for exception, FORCED_EXIT|val for SystemExit
+    if (ret == 1) {
+        return 0; // success
+    } else if (ret & PYEXEC_FORCED_EXIT) {
+        return ret; // SystemExit with exit value in lower 8 bits
+    } else {
+        return 1; // exception
+    }
 }
 
 static int do_str(const char *str) {
```

Comment:
It would be possible to replace the `execute_from_lexer()` below with `pyexec_vstr()` (just make a temporary vstr on the C stack and populate it with `str` and `strlen(str)`).

Might be worth doing that as part of this PR?

Reply:
`execute_from_lexer` is also still used for a couple of other things. I've trimmed it down a bit to remove now-unused functionality but would it be better refactored into a couple of new `pyexec_str_single()` and `pyexec_stdin()` functions in pyexec just to consolidate the location of all this sort of code? These functions are only needed on desktop really so would probably want extra #defines around them.

Reply:
I've done this as well now thanks